crates/ironrdp-pdu/src/lib.rs

```diff
@@ -128,7 +128,7 @@ pub trait Pdu {
 /// The resulting binary payload is a fully encoded PDU that may be sent to the peer.
 ///
 /// This trait is object-safe and may be used in a dynamic context.
-pub trait PduEncode {
+pub trait PduEncode: Send + Sync + fmt::Debug {
```

Comment:
These traits were needed for integration with Teleport's current system, they didn't break anything on IronRDP's side afaik so I figured they were acceptable.

Reply:
Hmm, this change seems harmless for PDUs, but maybe @CBenoit will have a different opinion on that?

Reply:
It’s more flexible to add the additional trait bounds at the place that really needs them and this is reducing a bit the number of types on which we may implement the trait, but I think this is quite reasonable for all PDUs to be `Send` + `Sync`. To be honest I’m not sure we actually want to enforce all PDUs to be `Debug`, but we derive this trait most of the time as well, so it’s probably not too bad either.
One concern I have is that one major way to reduce the size of our WASM artifacts would be to turn off the `Debug` derives in production builds, and by requiring the `Debug` trait this way we are probably making this harder for us in the future.

Mind if I ask exactly how this solves the issue with Teleport’s current system?

Reply:
> Mind if I ask exactly how this solves the issue you have with Teleport’s current system?

`Send` + `Sync` are required for Teleport for a couple of reasons:

1. As written, so that we can use `RdpdrPdu` in a `Send` + `Sync` manner.
2. So that `SvcProcessorMessages`, which contain a `Vec<SvcMessage>` which contains a `Box<dyn PduEncode>`, can be used in a `Send` + `Sync` manner.

`Debug` was added because:

`Box<dyn PduEncode>` is part of `DeviceControlResponse`:

https://github.com/Devolutions/IronRDP/blob/1a98d69d9a7652e59d55c136b18233dd7da7ad88/crates/ironrdp-rdpdr/src/pdu/efs.rs#L1154-L1160

and `DeviceControlResponse` is part of `RdpdrPdu`:

https://github.com/Devolutions/IronRDP/blob/1a98d69d9a7652e59d55c136b18233dd7da7ad88/crates/ironrdp-rdpdr/src/pdu/mod.rs#L15-L26

and `RdpdrPdu` has a custom `Debug` implementation which I've needed to verify that these intermediate implementation steps are working correctly

https://github.com/Devolutions/IronRDP/blob/1a98d69d9a7652e59d55c136b18233dd7da7ad88/crates/ironrdp-rdpdr/src/pdu/mod.rs#L144-L173

Currently the only thing that goes into that `Box<dyn Encode>` is the `LongReturn`, however soon there will be a variety of other structures. However I can make another trait which has these constraints rather than using PduEncode so that I can rid us of the Debug:

0a9fbf06d3de1f01dcd202fffc0cf9f94606b12f

Reply:
> `Send` + `Sync` are required for Teleport for a couple of reasons:
> 
>     1. As written, so that we can use `RdpdrPdu` in a `Send` + `Sync` manner.
> 
>     2. So that `SvcProcessorMessages`, which contain a `Vec<SvcMessage>` which contains a `Box<dyn PduEncode>`, can be used in a `Send` + `Sync` manner.

Okay I see! I was able to spot that in your code on Teleport side and I now understand better.

Note that in that case, a less restrictive approach is to add this bound at [the place where this is required directly](https://github.com/Devolutions/IronRDP/blob/d8c6efa3d180110812e2f6b28c9313981b3c020b/crates/ironrdp-svc/src/lib.rs#L60). Also, and I might be wrong, but it looks like only the `Send` bound is required:
```rust
pdu: Box<dyn PduEncode + Send>,
```

That being said, a PDU structure that is not `Sync` (non thread-safe interior mutability) or even non `Send` (non thread-safe ref counting) would be pretty weird in my opinion, so I’m not really against leaving it as-is either.

I’ll leave you decide on that!

> `Debug` was added because […]

Fair enough, yes I agree with your approach of using another trait internally in that case

Reply:
> Note that in that case, a less restrictive approach is to add this bound at [the place where it is required directly](https://github.com/Devolutions/IronRDP/blob/d8c6efa3d180110812e2f6b28c9313981b3c020b/crates/ironrdp-svc/src/lib.rs#L60).

It makes sense generally to be as conservative as possible with constraints in order to keep the trait maximally flexible. I agree that a `!Send` or `!Sync` `PduEncode` seems odd, but will avoid arbitrarily restricting us/others on principle.

Initially I opted for constraining `PduEncode` because `pdu: Box<dyn PduEncode + Send + Sync + std::fmt::Debug>` isn't allowed:

> > only auto traits can be used as additional traits in a trait object
consider creating a new trait with all of these as supertraits and using that trait here instead: `trait NewTrait: PduEncode + Debug {}`
auto-traits like `Send` and `Sync` are traits that have special properties; for more information on them, visit <https://doc.rust-lang.org/reference/special-types-and-traits.html#auto-traits>

> Also, and I might be wrong, it looks like only the `Send` bound is required

That's correct, I was confused. Again I'll opt for maximal conservatism.

1b31b3625ffc1993a8a533a9f2a951bd1dff5ec3